umbrella-feeds-reader/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+//! # Umbrella feeds
+//!
+//! This is an optional price reader for just one feed.
+//! For maximum gas optimization it is recommended to use UmbrellaFeeds directly.
+//!
+//! This contract has a hard-coded `umbrella_feed` contract address.
+//! ATTENTION: `Umbrella_feed` contract should only be upgraded via the `atomicUpdate` entry point in the registry contract to not break this link.
```

Comment:
if `Umbrella_feed` is upgradable, then do we ever need to call `atomicUpdate` on concordium? It will be the same address right?

Reply:
Good point. Unless you want a way to recover from incorrect upgrades of the feed contract. For example, if you accidentally upgrade it to a contract that cannot be further upgraded. I see that Doris added a comment in the feed contract about always keeping the upgrade entrypoint.

Reply:
To natively upgrade a smart contract on Concordium, it has to have this line in its logic:

```
    // Trigger the upgrade.
    host.upgrade(param.module)?;
```

If you call `atomicUpdate` on the registry to upgrade the `Umbrella_feeds` contract, it triggers the above line. The `Umbrella_feeds` contract address will stay the same but the pointer to the module reference (that includes the logic of the contract), will change to the new module reference provided. So you need to call the `atomicUpdate` entry point to upgrade the contract logic and currently, this is done via the registry to have one owner (the owner of the registry) that is responsible for all upgrades/replacements of contract addresses in the whole protocol.

@Bargsteen I think there is no good way to recover from incorrect upgrades :). The protocol is broken and stays broken and that is problematic for immutable contracts that listen to these contracts for price information. So make sure you do upgrades with care. I added comments that you should not break the upgrade mechanism accidentally by e.g. forgetting to include the hooks function (`unregister()` and `upgradeNatively()`) in the Umbrella feeds contract. 

@DZariusz Looking at the solidity contracts, it seems to me that the `registry` and `staking_bank` as well as `umbrell_feeds_reader` are all contracts that can never be replaced/updated without breaking the old protocol (meaning breaking protocols that are immutable and listen to these contracts for price information). Based on this assumption, I also only made the `Umbrella_feeds` contract upgradable. If this assumption is wrong then please let me know how this replacement mechanism on solidity was intended exactly.

Reply:
this is the flow for updating contract that we have:
1. `Registry` is the one that will be never updated. 
2. `StakingBank` can be updated any time eg when we have new validators, we are deploying new contract and then we calling `importContracts`. We can also call `atomicUpdate` but for bank it does not matter, because bank not using register/unregister "hooks". Importing new address into registry does not break protocol (and it should not). If users will call eg `UmbrellaFeeds` for a data they will not get any reverts at any point.
    - notice: `StakingBank` address is immutable in other contracts so they will still verify signatures based on old data, so it is important to simply redeploy them (so immutable address can be set to new one). Otherwise contract will not be in sync with validators list, but it will not break protocol for users.
3. `UmbrellaFeeds` - this contract has `destroy()` method that can be called if new contract is registered. So if users not using registry, it will break for them. If they are suing (they should) they it will not break protocol at any point.
4. `UmbrellaFeedsReader` - this contract is not registered in `Registry`. However it uses `Registry` to do fallback call in case `UmbrellaFeeds` for which it was deployed, not longer exists (was destroyed). So there is no scenario when it breaks.

summing up - there should be no scenario (if users are using protocol in a correct way, so via registry)  in which protocol will break for them.

In case of concordium, If there is a chance that upgrade can go wrong, then we should also "force" users to use registry. **And registry should not be upgradable**.

Let's do a call if necessary.